File: tests/angle.py
#!/usr/bin/env python

# 
# LSST Data Management System
# Copyright 2008, 2009, 2010 LSST Corporation.
# 
# This product includes software developed by the
# LSST Project (http://www.lsst.org/).
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the LSST License Statement and 
# the GNU General Public License along with this program.  If not, 
# see <http://www.lsstcorp.org/LegalNotices/>.
#

"""
Tests for Angle

Run with:
   angle.py
or
   python
   >>> import angle; angle.run()
"""

import math, os, sys
import unittest
import lsst.utils.tests as tests

import lsst.afw.geom as afwGeom

#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

class AngleTestCase(unittest.TestCase):
    """A test case for Angle"""
    def setUp(self):
        self.pi = afwGeom.Angle(math.pi, afwGeom.radians)
        self.d = 180*afwGeom.degrees

    def tearDown(self):
        pass

    def testCtor(self):
        self.assertEqual(self.pi, math.pi)
        self.assertEqual(self.pi, afwGeom.Angle(math.pi))
        self.assertEqual(self.pi, self.d)

        dd = afwGeom.Angle(180, afwGeom.degrees)
        self.assertEqual(self.d, dd)
        dd = afwGeom.Angle(60*180, afwGeom.arcminutes)
        self.assertEqual(self.d, dd)
        dd = afwGeom.Angle(60*60*180, afwGeom.arcseconds)
        self.assertEqual(self.d, dd)

    def testArithmetic(self):
        self.assertTrue(afwGeom.isAngle(self.pi))
        self.assertFalse(afwGeom.isAngle(self.pi.asRadians()))
        self.assertFalse(afwGeom.isAngle(math.pi))
        
        def tst():
            self.pi - math.pi           # subtracting a float from an Angle
        self.assertRaises(TypeError, tst)
        self.assertEqual(self.pi - math.pi*afwGeom.radians, 0) # OK with units specified
        self.assertEqual(self.pi - self.d, 0)                  # can subtract Angles

        def tst():
            self.pi + math.pi           # adding a float to an Angle
        self.assertRaises(TypeError, tst)

        def tst():
            self.pi*afwGeom.degrees     # self.pi is already an Angle
        self.assertRaises(NotImplementedError, tst)

        self.assertEqual((self.pi + self.d).asAngularUnits(afwGeom.degrees), 360)
        self.assertEqual((self.pi).asRadians(), math.pi)
        self.assertEqual((self.pi/2).asDegrees(), 90)
        self.assertEqual((self.pi*2).asArcminutes(), 360*60)
        self.assertEqual((self.pi*2).asArcseconds(), 360*60*60)

        self.assertEqual(math.sin(self.pi/2), 1.0) # automatic conversion to double

    def testPi(self):
        self.assertEqual(afwGeom.PI, math.pi)

    def testComparison(self):
        a2 = 2.0 * afwGeom.arcseconds
        a1 = 0.5 * afwGeom.arcseconds
        a3 = 0.5 * afwGeom.arcseconds
        print 'a1', a1
        print 'a2', a2
        print 'a3', a3
        self.assertEqual(a1 == a3, True)
        self.assertEqual(a1 != a2, True)
        self.assertEqual(a1 <= a2, True)
        self.assertEqual(a1 <  a2, True)
        self.assertEqual(a2 >  a1, True)
        self.assertEqual(a2 >= a1, True)

        self.assertEqual(a1 != a3, False)
        self.assertEqual(a1 == a2, False)
        self.assertEqual(a1 >= a2, False)
        self.assertEqual(a1 >  a2, False)
        self.assertEqual(a2 <  a1, False)
        self.assertEqual(a2 <= a1, False)

    def testTrig(self):
        self.assertEqual(math.cos(self.d), -1.0)
        self.assertAlmostEqual(math.sin(self.d),  0.0, places=15)
        thirty = 30.*afwGeom.degrees
        self.assertAlmostEqual(math.sin(thirty), 0.5, places=15)
        


#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

def suite():
    """Returns a suite containing all the test cases in this module."""
    tests.init()

    suites = []
    suites += unittest.makeSuite(AngleTestCase)
    suites += unittest.makeSuite(tests.MemoryTestCase)
    return unittest.TestSuite(suites)

def run(shouldExit=False):
    """Run the tests"""
    tests.run(suite(), shouldExit)

if __name__ == "__main__":
    run(True)

File: python/lsst/afw/geom/angle.cc
/*
 * LSST Data Management System
 * Copyright 2008-2016  AURA/LSST.
 *
 * This product includes software developed by the
 * LSST Project (http://www.lsst.org/).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the LSST License Statement and
 * the GNU General Public License along with this program.  If not,
 * see <https://www.lsstcorp.org/LegalNotices/>.
 */

#include <pybind11/pybind11.h>
#include <pybind11/operators.h>

#include "lsst/afw/geom/Angle.h"

namespace py = pybind11;

using namespace lsst::afw::geom;

PYBIND11_PLUGIN(_angle) {
    py::module mod("_angle", "Python wrapper for afw _angle library");

    py::class_<AngleUnit> clsAngleUnit(mod, "AngleUnit");

    /* Operators */
    clsAngleUnit.def(py::self == py::self);
    mod.def("AngleUnit_mul", [](lsst::afw::geom::AngleUnit const& lhs, double rhs) -> Angle {
        return rhs*lhs;
    });
    mod.def("AngleUnit_mul", [](double lhs, lsst::afw::geom::AngleUnit const& rhs) -> Angle {
        return lhs*rhs;
    });

    /* Members */
    mod.attr("radians") = py::cast(radians);
    mod.attr("degrees") = py::cast(degrees);
    mod.attr("hours") = py::cast(hours);
    mod.attr("arcminutes") = py::cast(arcminutes);
    mod.attr("arcseconds") = py::cast(arcseconds);

    py::class_<Angle> clsAngle(mod, "Angle");

    /* Constructors */
    clsAngle.def(py::init<double, AngleUnit>(), py::arg("val"), py::arg("units")=radians);
    clsAngle.def(py::init<>());

    /* Operators */
    clsAngle.def(py::self *= float());
    clsAngle.def(py::self *= int());
    clsAngle.def(py::self += float());
    clsAngle.def(py::self += int());
    clsAngle.def(py::self -= float());
    clsAngle.def(py::self -= int());
    clsAngle.def(py::self == py::self);
    clsAngle.def(py::self != py::self);
    clsAngle.def(py::self <= py::self);
    clsAngle.def(py::self >= py::self);
    clsAngle.def(py::self < py::self);
    clsAngle.def(py::self > py::self);
    clsAngle.def(py::self - py::self);
    clsAngle.def(py::self + py::self);
    clsAngle.def("__float__", &Angle::operator double);
    clsAngle.def("__div__", [](Angle &a, double d) {
        return a / d;
    });
    clsAngle.def("__rdiv__", [](Angle &a, double d) {
        return a / d;
    });
    mod.def("Angle_mul", [](Angle &lhs, Angle &rhs) {
        return lhs * rhs;
    });
    mod.def("Angle_mul", [](Angle &lhs, double rhs) {
        return lhs * rhs;
    });
    mod.def("Angle_mul", [](Angle &lhs, int rhs) {
        return lhs * rhs;
    });

    /* Members */
    clsAngle.def("asAngularUnits", &Angle::asAngularUnits);
    clsAngle.def("asRadians", &Angle::asRadians);
    clsAngle.def("asDegrees", &Angle::asDegrees);
    clsAngle.def("asHours", &Angle::asHours);
    clsAngle.def("asArcminutes", &Angle::asArcminutes);
    clsAngle.def("asArcseconds", &Angle::asArcseconds);
    clsAngle.def("toUnitSphereDistanceSquared", &Angle::toUnitSphereDistanceSquared);
    clsAngle.def("fromUnitSphereDistanceSquared", &Angle::fromUnitSphereDistanceSquared);
    clsAngle.def("wrap", &Angle::wrap);
    clsAngle.def("wrapCtr", &Angle::wrapCtr);
    clsAngle.def("wrapNear", &Angle::wrapNear);

    /* Non members */
    mod.attr("PI") = py::float_(PI);
    mod.attr("TWOPI") = py::float_(TWOPI);
    mod.attr("HALFPI") = py::float_(HALFPI);
    mod.attr("ONE_OVER_PI") = py::float_(ONE_OVER_PI);
    mod.attr("SQRTPI") = py::float_(SQRTPI);
    mod.attr("INVSQRTPI") = py::float_(INVSQRTPI);
    mod.attr("ROOT2") = py::float_(ROOT2);
    mod.attr("NullAngle") = py::cast(NullAngle);

    mod.def("degToRad", degToRad);
    mod.def("radToDeg", radToDeg);
    mod.def("radToArcsec", radToArcsec);
    mod.def("radToMas", radToMas);
    mod.def("arcsecToRad", arcsecToRad);
    mod.def("masToRad", masToRad);
    mod.def("isAngle", (bool (*)(Angle const&)) isAngle);
    mod.def("isAngle", isAngle<double>);

    return mod.ptr();
}
